go/worker/compute/txnscheduler/committee/node.go

```diff
@@ -139,9 +141,76 @@ func (n *Node) getMetricLabels() prometheus.Labels {
 
 // HandlePeerMessage implements NodeHooks.
 func (n *Node) HandlePeerMessage(ctx context.Context, message *p2p.Message) (bool, error) {
+	if message.TxnScheduleTx != nil {
+		n.logger.Debug("received peer message", "message", message)
+
+		expectedEpochNumber := message.TxnScheduleTx.ExpectedEpochNumber
+		call := message.TxnScheduleTx.Data
+
+		// Check if we are a leader. Note that we may be in the middle of a
+		// transition, but this shouldn't matter as the client will retry.
+		epochSnapshot := n.commonNode.Group.GetEpochSnapshot()
+		n.commonNode.CrossNode.Lock()
+		height := n.commonNode.CurrentHeight
+		n.commonNode.CrossNode.Unlock()
+		if !epochSnapshot.IsTransactionSchedulerLeader(height) {
+			n.logger.Error("unable to handle message, not leader anymore",
+				"err", api.ErrNotLeader,
+				"current_epoch", epochSnapshot.GetEpochNumber(),
+				"current_height", height,
+			)
+			return true, p2pError.Permanent(api.ErrNotLeader)
+		}
+
+		// Check if expected client's epoch matches the current worker's one.
+		// TODO: maybe not needed anymore now that we check group version in
+		// messages.
+		if epochSnapshot.GetEpochNumber() != expectedEpochNumber {
+			n.logger.Error("unable to handle message",
+				"err", api.ErrEpochNumberMismatch,
+				"current_epoch_number", epochSnapshot.GetEpochNumber(),
+				"expected_epoch_number", expectedEpochNumber,
+			)
+			return true, p2pError.Permanent(api.ErrEpochNumberMismatch)
+		}
+
+		if n.checkTxEnabled {
+			// Check transaction before queuing it.
+			if err := n.CheckTx(ctx, call); err != nil {
+				return true, p2pError.Permanent(err)
+			}
+			n.logger.Debug("worker CheckTx successful, queuing transaction")
+		}
+
+		err := n.QueueCall(call)
+		if err != nil {
+			n.logger.Error("unable to Queue call",
+				"err", err,
+			)
+		}
+		n.logger.Debug("message handeled", "message", message)
+		return true, err
+	}
 	return false, nil
 }
 
+// QueueCall queues a runtime call.
+func (n *Node) QueueCall(call []byte) error {
+	n.algorithmMutex.RLock()
+	defer n.algorithmMutex.RUnlock()
+
+	if n.algorithm == nil || !n.algorithm.IsInitialized() {
+		return api.ErrNotReady
+	}
+	if err := n.algorithm.ScheduleTx(call); err != nil {
+		return err
```

Comment:
Failing to queue (e.g., due to already being in the queue) should prevent relaying (e.g., a permanent P2P error). Currently I see that the batching algorithm does not propagate the error, but since there are no more client acks, it would be better to just return an error and make it permanent here.